proposals/0000-binaryinteger-ispower.md

```diff
@@ -0,0 +1,271 @@
+    var found = false
+    for word in words {
+      if word != 0 {
+        if found || word & (word - 1) != 0 { return false }
+        found = true
+      }
+    }
+    return found
+  }
+}
+```
+
+### Fast path when base itself is a power of two
+
+As long as we have `_isPowerOfTwo` available, it becomes easy to implement such fast path when the base itself is power of two.  It takes advantages of the existing APIs `isMultiple(of:)` and `trailingZeroBitCount`. The code can be written as below.
+
+```swift
+extension BinaryInteger {
+  @inlinable internal func _isPowerOf(powerOfTwo base: Self) -> Bool {
+    _precondition(base._isPowerOfTwo)
+    guard self._isPowerOfTwo else { return false }
+    return self.trailingZeroBitCount.isMultiple(of: base.trailingZeroBitCount)
+  }
+}
+```
+
+### Fast path when base is ten
+
+Unfortunately, there isn't any known super efficient algorithm to test if an integer is power of ten.  Some optimizations are discussed in the forum to help boost the performance to some extent.  Since the optimizations are conceptually non-trivial, they are not described here. Please refer to the [pitch thread](https://forums.swift.org/t/adding-ispowerof2-to-binaryinteger/24087/38) for details.
+
+
+### Slow path
+
+The slow path is generic for any input base, and the algorithm is standard. It handles some corner cases when `base.magnitude <= 1` in the first place; and then it repeatedly multiplies `base` to see if it can be equal to `self` at some point.  Attentions are required to avoid overflow of the multiplications.
+
+```swift
+extension BinaryInteger {
+  @usableFromInline internal func _slowIsPower(of base: Self) -> Bool {
+    // If self is 1, return true.
+    if self == 1 { return true }
+
+    // Here if base is 0, 1 or -1, return true iff self equals base.
+    if base.magnitude <= 1 { return self == base }
+
+    // At this point, we have base.magnitude >= 2. Repeatedly perform
+    // multiplication by a factor of base, and check if it can equal self.
+    let (bound, remainder) = self.quotientAndRemainder(dividingBy: base)
+    guard remainder == 0 else { return false }
+    var x: Self = 1
+    while x.magnitude < bound.magnitude { x *= base }
+    return x == bound
+  }
+}
+```
+
+## Source compatibility
+
+The proposed solution is an additive change.
+
+## Effect on ABI stability
+
+The proposed solution is an additive change.
+
+
+## Effect on API resilience
+
+The proposed solution is an additive change.
+
+
+## Alternatives considered
+
+### `isPower(of:)` as a protocol requirement
+
+Making the public API as a protocol requirement instead of an extension methods reserves future flexibility.  It can give users a chance to provide their own optimization on some custom types where the default implementation is inefficient.  However, it would increase the interface complexity of the heavily-used `BinaryInteger` protocol, so it may not be worthy.
```

Comment:
Remove the article `an`

```suggestion
Making the public API as a protocol requirement instead of extension methods reserves future flexibility.  It can give users a chance to provide their own optimization on some custom types where the default implementation is inefficient.  However, it would increase the interface complexity of the heavily-used `BinaryInteger` protocol, so it may not be worthy.
```